blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md

```diff
@@ -0,0 +1,189 @@
+for this is coming soon. I will also add support for UART-over-USB, which will
+allow controlling device using USB only.
+
+I have recently added basic LCD support (available on `display_support` branch),
+so you should be able to see UEFI booting without using UART. note that you
+won't be able to boot into OS, as this currently requires manual FDT loading.
+I will fix this problem in the near future.
+
+![UEFI boot menu](/img/xw711_uefi_screen.jpg)
+
+![UEFI grub](/img/xw711_uefi_grub.jpg)
+
+## Building and booting UEFI
+
+You can get BSP source code from 
+[my GitHub repo](https://github.com/arturkow2000/SunxiPlatformPkg). It has been
+tested with EDK II v2021.02 which can obtain from
+[TianoCore GitHub repo](https://github.com/tianocore/edk2/).
+
+```
+git clone --depth=1 https://github.com/tianocore/edk2 --branch edk2-stable202102
+cd edk2
+git submodule update --init --recursive
+git submodule add https://github.com/arturkow2000/SunxiPlatformPkg
+```
+
+If you want LCD support, you need to switch branch:
+
+```
+git checkout display_support
+```
+
+You can build it directly on your host system. Note that this has been tested
+on Ubuntu 20.04 with GCC 9.2.1, and it may not work on other distros with either
+too old or too new compiler.
+
+```
+make -C BaseTools/Source/C
+. edksetup.sh
+env GCC5_ARM_PREFIX=arm-none-eabi- build -a ARM -p SunxiPlatformPkg/XW711.dsc -t GCC5
+```
+
+Alternatively you can use 3mdeb Docker container that provides stable build
+environment independent of your distro.
+
+```
+$ docker run --rm -it -w /home/edk2/edk2 -v $PWD:/home/edk2/edk2 3mdeb/edk2 /bin/bash
+(docker)$ make -C BaseTools
+(docker)$ . edksetup.sh
+(docker)$ env GCC5_ARM_PREFIX=arm-linux-gnueabihf- build -a ARM -p SunxiPlatformPkg/XW711.dsc -t GCC5
+```
+
+This builds two files: `SUNXI_SPL.fd` and `SUNXI_EFI.fd`. For `sunxi-fel` to
+accept SPL, you have to patch its header. Tool for this purpose is written in
+Rust, so if you don't have Rust installed already, you can get it from
+https://rustup.rs.
+
+```
+cd Build/XW711/DEBUG_GCC5/FV/
+cargo install --git https://github.com/arturkow2000/sunxiboot
+sunxiboot checksum SUNXI_SPL.fd
+sunxi-fel --verbose spl SUNXI_SPL.fd write 0x42000000 SUNXI_EFI.fd exe 0x42000000
+```
+
+[![asciicast](https://asciinema.org/a/pCr0fQKHnBFRhHyWjC9ml3Gcz.svg)](https://asciinema.org/a/pCr0fQKHnBFRhHyWjC9ml3Gcz?speed=1)
+
+## UEFI in action
+
+This video shows booting EDK II and Fedora ARM with FDT, few workarounds are
+required:
+
+- DTB must be manually loaded from file on each boot, using `setfdt` EFI shell
+  command
+
+- I had to boot Linux with `cpufreq.off=1`, or it would hang while trying to
+  raise CPU frequency (since there is no frequency scaling, CPU is left at 384
+  MHz), this looks like Linux's bug.
+
+- I had to disable `axp20x_adc` driver because it was causing board to power off
+  instantly, this isn't related to EFI itself, yet still, it's causing problems.
```

Comment:
Let's be consistent about naming UEFI is the name of specification. EFI is pre-UEFI non-public specification, so I doubt you using it. To be precise we can use UEFI firmware or just firmware, boot firmware even EDK II.